### PR TITLE
Multiple areas

### LINKED ISSUE #5: Improve the visualization of differences
Improve the visualization of differences, separating each point into its own areas and discriminating each one with its own rectangles.

### PR DESCRIPTION
**Pull request by cristianofmc**
Code update for better discrimination of points of difference, separating each point of difference with its own rectangle.
image
image
image

- rake spec
image
- rake rubocop
image
- rake rubocop:md
image

### CODE REVIEW

File: lib/image_compare/modes/base.rb [hunk 1]
```diff
@@ -6,22 +6,22 @@ class Base
...
+        @different_areas = []
       end
 
       def compare(a, b)
         result.image = a
         @include_rect ||= a.bounding_rect
-        @bounds = Rectangle.new(*include_rect.bounds)
-
         b.compare_each_pixel(a, area: include_rect) do |b_pixel, a_pixel, x, y|
+
```
Comment (gsguma) [inline lib/image_compare/modes/base.rb:39]: Consider removing this extra line

File: lib/image_compare/modes/base.rb [hunk 2]
```diff
@@ -36,9 +36,15 @@ def diff(bg, diff)
         diff.each do |pixels_pair|
           pixels_diff(diff_image, *pixels_pair)
         end
-        create_diff_image(bg, diff_image)
-          .highlight_rectangle(bounds)
-          .highlight_rectangle(include_rect, :green)
+
+        diff_image = create_diff_image(bg, diff_image)
+        if @different_areas.any?
```
Comment (gsguma) [inline lib/image_compare/modes/base.rb]: I think this validation is redundant (_the each method already ignores empty arrays by default_)

File: lib/image_compare/rectangle.rb [hunk 3]
```diff
@@ -31,5 +31,40 @@ def contains_point?(x, y)
         (x <= right && y <= top && x >= left && y >= bot) ||
         (x.between?(right, left) && y.between?(bot, top))
     end
+
+    def close_to_the_area?(x, y)
+      offsets = [-1, 0, 1]
+
```
Comment (gsguma) [inline lib/image_compare/rectangle.rb]: Consider remove this "extra line"

File: lib/image_compare/rectangle.rb [hunk 4]
```diff
@@ -31,5 +31,40 @@ def contains_point?(x, y)
...
+      offsets.product(offsets).any? do |dx, dy|
+        next if dx.zero? && dy.zero?
+        contains_point?(x + dx, y + dy)
+      end
+    end
+
+    def rect_close_to_the_area?(rect)
+      offsets = [-1, 0, 1]
+
```
Comment (gsguma) [inline lib/image_compare/rectangle.rb]: Consider remove this "extra line"

File: spec/base_spec.rb [hunk 5]
```diff
@@ -0,0 +1,148 @@
+require "spec_helper"
+require "image_compare/modes/base"
```
Comment (gsguma) [inline spec/base_spec.rb]: Consider using single quotes for non-interpolated strings

File: spec/base_spec.rb [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+describe ImageCompare::Modes::Base do
+  let(:base) { described_class.new(threshold: 0.5, lower_threshold: 0.2) }
+  let(:image_a) { double("Image") }
```
Comment (gsguma) [inline spec/base_spec.rb:6]: Consider using single quotes for non-interpolated strings

File: spec/base_spec.rb [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+  let(:image_b) { double("Image") }
```
Comment (gsguma) [inline spec/base_spec.rb]: Consider using single quotes for non-interpolated strings

File: spec/integrations/color_spec.rb [hunk 8]
```diff
@@ -45,6 +45,7 @@
     end
 
     it "creates correct difference image" do
+      subject.difference_image.save("color_diff.png")
```
Comment (gsguma) [inline spec/integrations/color_spec.rb:48]: Consider using single quotes for non-interpolated strings

File: spec/integrations/color_spec.rb [hunk 9]
```diff
@@ -45,6 +45,7 @@
...
       expect(subject.difference_image).to eq(ImageCompare::Image.from_file(image_path("color_diff")))
```
Comment (gsguma) [inline spec/integrations/color_spec.rb:49]: Consider using single quotes for non-interpolated strings

File: spec/integrations/delta_spec.rb [hunk 10]
```diff
@@ -45,6 +45,7 @@
     end
 
     it "creates correct difference image" do
+      subject.difference_image.save("delta_diff.png")
       expect(subject.difference_image).to eq(ImageCompare::Image.from_file(image_path("delta_diff")))
```
Comment (gsguma) [inline spec/integrations/delta_spec.rb]: Consider using single quotes for non-interpolated strings

File: spec/integrations/grayscale_spec.rb [hunk 11]
```diff
@@ -23,6 +23,7 @@
     end
 
     it "creates correct difference image" do
+      subject.difference_image.save("grayscale_diff.png")
       expect(subject.difference_image).to eq(ImageCompare::Image.from_file(image_path("grayscale_diff")))
```
Comment (gsguma) [inline spec/integrations/grayscale_spec.rb]: Consider using single quotes for non-interpolated strings

File: spec/integrations/rgb_spec.rb [hunk 12]
```diff
@@ -23,6 +23,7 @@
     end
 
     it "creates correct difference image" do
+      subject.difference_image.save("rgb_diff.png")
       expect(subject.difference_image).to eq(ImageCompare::Image.from_file(image_path("rgb_diff")))
```
Comment (gsguma) [inline spec/integrations/rgb_spec.rb]: Consider using single quotes for non-interpolated strings

**Review by gsguma — DISMISSED** [review]
Good job! 💯

**Comment by cristianofmc** [comment]
image

File: lib/image_compare/modes/base.rb [hunk 13]
```diff
@@ -6,11 +6,11 @@ class Base
...
       include ColorMethods
 
-      attr_reader :result, :threshold, :lower_threshold, :different_areas, :exclude_rect, :include_rect
+      attr_reader :result, :threshold, :lower_threshold, :different_areas, :exclude_rects, :include_rect
 
-      def initialize(threshold: 0.0, lower_threshold: 0.0, exclude_rect: nil, include_rect: nil)
+      def initialize(threshold: 0.0, lower_threshold: 0.0, exclude_rects: [], include_rect: nil)
         @include_rect = Rectangle.new(*include_rect) unless include_rect.nil?
-        @exclude_rect = Rectangle.new(*exclude_rect) unless exclude_rect.nil?
+        @exclude_rects = @exclude_rects = exclude_rects.empty? ? Set.new : Set.new(exclude_rects.map { |rect| Rectangle.new(*rect) })
```
Comment (gsguma) [inline lib/image_compare/modes/base.rb]: Please consider changing to:
```
@exclude_rects = exclude_rects.empty? ? Set.new : Set.new(exclude_rects.map { |rect| Rectangle.new(*rect) })
```

_**Don't forget to fix the error on line 17 too**_

File: lib/image_compare/modes/base.rb [hunk 14]
```diff
@@ -94,9 +94,9 @@ def update_bounds(x, y)
       end
 
       def area
-        area = include_rect.area
-        return area if exclude_rect.nil?
-        area - exclude_rect.area
+        total_area = @include_rect.area
+        total_exclude_area = @exclude_rects.nil? ? 0 : @exclude_rects.sum { |rect| rect.area }
```
Comment (gsguma) [inline lib/image_compare/modes/base.rb]: Please consider changing to `total_exclude_area = @exclude_rects&.sum(&:area) || 0`

File: spec/base_spec.rb [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+        result = ImageCompare.compare(path_3, path_4, **options)
+        result.difference_image.save('multiple_exclude_rects_with_multiple_differences.png')
+
+        expect(result).to be_a ImageCompare::Result
+        expect(result.send(:score)).to be > 0 and be < 1
+      end
+    end
+
+  end
+end
```
Comment (gsguma) [inline spec/base_spec.rb]: Missing EOF

**Review by gsguma — CHANGES_REQUESTED** [review]

File: spec/base_spec.rb [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+
+      it "compares two images and returns a Result object" do
+        result = ImageCompare.compare(path_3, path_4, **options)
+        result.difference_image.save('multiple_exclude_rects_with_multiple_differences.png')
+
+        expect(result).to be_a ImageCompare::Result
+        expect(result.send(:score)).to be > 0 and be < 1
+      end
+    end
+
```
Comment (gsguma) [inline spec/base_spec.rb]: Please remove extra line

**Review by gsguma — CHANGES_REQUESTED** [review]

**Comment by gsguma** [comment]
Don't forget to inspect code quality, running:
```
rake rubocop
rake rubocop:md
```

**Review by gsguma — DISMISSED** [review]

**Review by gsguma — APPROVED** [review]

**Pull request merged**